circuits/semaphore-base.circom

```diff
@@ -0,0 +1,90 @@
+pragma circom 2.0.0;
+
+include "../node_modules/circomlib/circuits/poseidon.circom";
+include "./tree.circom";
+
+template CalculateSecret() {
+    signal input identity_nullifier;
+    signal input identity_trapdoor;
+
+    signal output out;
+
+    component hasher = Poseidon(2);
+    hasher.inputs[0] <== identity_nullifier;
+    hasher.inputs[1] <== identity_trapdoor;
+    out <== hasher.out;
+}
+
+template CalculateIdentityCommitment() {
+    signal input secret_hash;
+
+    signal output out;
+
+    component hasher = Poseidon(1);
+    hasher.inputs[0] <== secret_hash;
+    out <== hasher.out;
+}
+
+template CalculateNullifierHash() {
+    signal input external_nullifier;
+    signal input identity_nullifier;
+
+    signal output out;
+
+    component hasher = Poseidon(2);
+    hasher.inputs[0] <== external_nullifier;
+    hasher.inputs[1] <== identity_nullifier;
+    out <== hasher.out;
+}
+
+// n_levels must be < 32
+template Semaphore(n_levels) {
+
+    var LEAVES_PER_NODE = 5;
+    var LEAVES_PER_PATH_LEVEL = LEAVES_PER_NODE - 1;
+
+    signal input signal_hash;
+    signal input external_nullifier;
+
+    signal input identity_nullifier;
+    signal input identity_trapdoor;
+    signal input identity_path_index[n_levels];
+    signal input path_elements[n_levels][LEAVES_PER_PATH_LEVEL];
+
+    signal output root;
+    signal output nullifierHash;
+
+    component secret = CalculateSecret();
+    secret.identity_nullifier <== identity_nullifier;
+    secret.identity_trapdoor <== identity_trapdoor;
+
+    signal secret_hash;
+    secret_hash <== secret.out;
+
+    component identity_commitment = CalculateIdentityCommitment();
+    identity_commitment.secret_hash <== secret_hash;
+
+    component calculateNullifierHash = CalculateNullifierHash();
+    calculateNullifierHash.external_nullifier <== external_nullifier;
+    calculateNullifierHash.identity_nullifier <== identity_nullifier;
+
+    var i;
+    var j;
+    component inclusionProof = QuinTreeInclusionProof(n_levels);
+    inclusionProof.leaf <== identity_commitment.out;
+
+    for (i = 0; i < n_levels; i++) {
+      for (j = 0; j < LEAVES_PER_PATH_LEVEL; j++) {
+        inclusionProof.path_elements[i][j] <== path_elements[i][j];
+      }
+      inclusionProof.path_index[i] <== identity_path_index[i];
+    }
+
+    root <== inclusionProof.root;
+
+    // Dummy square to prevent tampering signalHash
+    signal signal_hash_squared;
+    signal_hash_squared <== signal_hash * signal_hash;
```

Comment:
What kind of tampering and how does this prevent it?

Does Circom not remove this as dead code since `signal_hash_squared` is unused?

Reply:
To answer my first question: `signal_hash` would otherwise be unused and we need the proof verification to fail if this public input changes.

Reply:
So my argument for this is that since signal is not used in circuit and let's say you are using some realyer, it can temper the signal but proof can still be valid. 

But yes, your answer is good!

Reply:
Here's a relevant snip from the circom doc.
> If a signal is not used in any constraint, an *error* message will be generated. Moreover, if such signal is an input signal `s` , then the compiler would suggest adding a constraint of the form `x * 0 === 0`. 
> [source](https://docs.circom.io/2.-circom-fundamentals/circuit-structure#compiler-messages)

They recommend a simpler constraint, but it looks like adding a dummy constraint is the recommended way to solve the issue of unused public inputs.